discussion/rust-for-embedded-systems.tex

```diff
@@ -0,0 +1,90 @@
+That being said, they do not necessarily use any functionality that we have not already made available through \gls{rel} in our project.
+This simple fact renders nearly all (if not all) packages available through \texttt{crates.io} unusable for our project, even though the same basis that many of these packages depend on might already be available through \gls{rel}.
+
+As an example to demonstrate this problem, we look to the \texttt{lazy-static}\footnote{\url{https://crates.io/crates/lazy_static}} project available on \texttt{crates.io}.
+This project allows the programmer to declare \emph{static} variables that get initialized at \emph{runtime}.
+It provides some of the same functionality that we implemented for the event-hub described in \autoref{sec:irq-closures}.
+We can not use this project directly because it depends on {\std}, but if we look closer at the project's source code, we can see that the project author has implemented an alternate (and outdated) version which is only dependent on {\core}.
+Another example is the \texttt{smallvec}\footnote{\url{https://crates.io/crates/smallvec}} project, which provides a handful of optimized versions of {\rust}'s vector structure that have a length-limit of 32 entries.
+Similar to the \texttt{lazy-static} project, all the dependencies for \texttt{smallvec} are included from {\std}, but are also available in {\core}.
+We have implemented a structure that is similar to \texttt{smallvec} as part of the {\tracker} application, but it feels like an unnecessary addition because it is already available on \texttt{crates.io} \emph{and} it only depends on features from {\core}.
+
+This distinction between {\std} and \gls{rsl} feels somewhat contradictory to {\rust}'s focus on modularization and package distribution with {\cargo}.
+It also feels destructive for non-standard projects like ours, because we would either have to re-implement or modify the already existing projects in order to get them to work for our platform.
+We hope that it will get easier in time to define non-standard projects, while simultaneously being able to utilize many of the great open-source libraries that are already available.
+
+\subsection{Language Challenges}
+\label{sec:disc:lang-challenges}
+
+In \autoref{sec:intro:assignment} we identified six language challenges that must be considered when using a language in an embedded system.
+Here, we revisit each challenge and discuss how they were solved for the {\rg} platform.
+
+\begin{description}
+\item [LC1 {\lci}] \hfill \\
+  {\rust} exposes two \concept{intrinsics} for handling volatile read and write.
+  These makes the code more verbose compared to the mechanism in {\C}, where a variable is marked with the \code{volatile} keyword.
+  In {\rust}, the intrinsic functions must be used each time a variable is read or written.
+  This, however, gives the programmer more fine-grained control as a variable can be used both as volatile and non-volatile.
```

Comment:
Append `An aspect we did not go into in this project is {\rust} ability to through traits redefine the semantics when using values, this concept can be applied to remove the verbosity of these operations.`